doc/explanation/package-management.md

```diff
@@ -0,0 +1,226 @@
+process by default will use the newest set of packages available.
+
+However, it is also possible to specify specific revisions of the repositories,
+to get a reproducible solution. Due to using Git, any previous revision of the
+repository can be used by specifying a commit hash.
+
+Dune defines two repositories by default:
+
+* `upstream` refers to the default branch of `opam-repository`, which
+  contains all the publically released packages.
+* `overlay` refers to
+  [opam-overlay](https://github.com/ocaml-dune/opam-overlays), which defines
+  packages patched to work with package management. The long-term goal is to
+  have as few packages as possible in this repository as more and more packages
+  work within Dune Package Management upstream. Check the
+  [compatibility](#compatibility) section for details.
+
+### Solving
+
+After Dune has retrieved the constraints and the set of possible packages, it is
+necessary to determine which packages and versions should be selected for the
+package lock.
+
+To do so, Dune uses
+[`opam-0install-solver`](https://github.com/ocaml-opam/opam-0install-solver),
+which is a variant of the `0install` solver to find solutions for opam packages.
+
+Contrary to opam, the Dune solver always starts from a blank slate. It
+assumes nothing is installed and everything needs to be installed. This has the
+advantage that solving is now simpler, and previous solver solutions don't interfere
+with the current one. Thus, given the same inputs, it should always have the same
+outputs; no state is held between the solver runs.
+
+This can lead to more packages being installed (as opam won't install new package versions
+by default if the existing versions satisfy the
+constraints), but it avoids interference from already installed packages that
+lead to potentially different solutions.
+
+After solving is done, the solution gets written into the lock directory with
+all the metadata necessary to build and install the packages. From this
+point on, there is no need to access any package repositories.
+
+:::{note}
+Solving and locking does not download the package sources. These are downloaded
+in a later step.
+:::
+
+## Building
+
+When building, Dune will read the information from the lock directory and set
+up rules to use the packages. Check {doc}`/explanation/mental-model` for
+details about rules.
+
+The rules that the package management sets up include:
+
+* Fetch rules to download the sources as well as any additional sources like
+  patches and unpack them
+* Build rules to evaluate the build instructions from the build instructions
+  stored in the lock directory
+* Install rules to put the artifacts that were built into the appropriate
+  Dune-managed folders
+
+Creating these processes as rules mean that they will only be executed on
+demand, so if the project has already downloaded the sources, it does not
+need to download them again. Likewise, if packages are installed, they stay
+installed.
```

Comment:
This is definitely the plan, however at the moment by default only downloading the dependencies is cached, thus I didn't want to introduce information that would require additional disclaimers (aka "this works but at the moment only if you manually set `(dune-cache enabled)` but we're working on it"). Thus I would wait with the inclusion of a paragraph like this until the feature lands as default and this would be true without any qualifiers.